sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeLocalShuffleReader.scala

```diff
@@ -139,6 +139,11 @@ object OptimizeLocalShuffleReader extends CustomShuffleReaderRule {
   def canUseLocalShuffleReader(plan: SparkPlan): Boolean = plan match {
     case s: ShuffleQueryStageExec =>
       s.mapStats.isDefined && supportLocalReader(s.shuffle)
+    case CustomShuffleReaderExec(s: ShuffleQueryStageExec, partitionSpecs)
+        if isUsedToCoalescePartitions(s.shuffle) =>
+      // Use LocalShuffleReader only when we can't CoalesceShufflePartitions
```

Comment:
Why? The general rule is to always use the local reader if possible. If coalesce happens first, local reader will respect the parallelism.